File: Data Structures & Algorithms/balanced-binary-tree/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isBalanced(root) {


        if (!root) return true;
        let leftTree = this.dfs(root.left);
        let rightTree = this.dfs(root.right);
        if (leftTree[0] && rightTree[0]) return Math.abs(leftTree[1] - rightTree[1]) <= 1;
        else return false;


    }

    dfs(root) {
        if (!root) return [true, 0];
        let [left, right] = [this.dfs(root.left), this.dfs(root.right)]
        const balanced = (Math.abs(left[1] - right[1]) <= 1) && left[0] && right[0];
        return [balanced, 1 + Math.max(left[1], right[1])];
    }

}

File: Data Structures & Algorithms/binary-tree-diameter/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        ans = 0
        def treeDepth(root):
            nonlocal ans
            if not root:
                return 0
            left = right = 0
            if root.left:
                left = treeDepth(root.left) + 1
            if root.right:
                right = treeDepth(root.right) + 1
            ans = max(ans, left+right)
            return max(left,right)
        treeDepth(root)
        return ans